proto/backend.proto

```diff
@@ -0,0 +1,212 @@
+syntax = "proto3";
+import "google/protobuf/empty.proto";
+package grafana.plugin;
+
+//-----------------------------------------------
+// Common
+//-----------------------------------------------
+
+message PluginConfig {
+  int64 id = 1;
+  int64 orgId = 2;
+  string name = 3;
+  string type = 4;
+  string url = 5;
+  string jsonData = 6;
+  map<string,string> decryptedSecureJsonData = 7;
+}
+
+
+//-----------------------------------------------
+// All Backend Plugins Implement
+//-----------------------------------------------
+
+service BackendPlugin {
+  // Get the plugin status
+  rpc Check(PluginStatusRequest) returns (PluginStatusResponse);
+
+  // HTTP Style request
+  rpc Resource(ResourceRequest) returns (ResourceResponse);
+
+  // Well typed query interface
+  rpc DataQuery(DataQueryRequest) returns (DataQueryResponse);
+
+}
+
+//-----------------------------------------------
+// Plugin Status Check
+//-----------------------------------------------
+
+message PluginStatusRequest {
+  // Plugin Configuration
+  PluginConfig config = 1;
+
+  // Environment info
+  map<string,string> headers = 2;
+
+  enum FetchInfo {
+    STATUS = 0;  // nothing
+    API = 1;     // OpenAPI description
+    METRICS = 2; // Prometheus style metrics
+    DEBUG = 3;   // JSON Debug Info (admin+ view)
+  }
+
+  // Get an OpenAPI description (viewer+)
+  FetchInfo fetch = 3;
+}
+
+message PluginStatusResponse {
+  enum PluginStatus {
+    UNKNOWN = 0; // starting etc
+    OK = 1;
+    ERROR = 2;
+  }
+
+  PluginStatus status = 1;
+  string info = 2;
+}
+
+//-----------------------------------------------
+// Data Query
+//-----------------------------------------------
+
+message TimeRange {
```

Comment:
I am thinking Two string fields (From and To). As well as an enum for the format of the two and from fields.  This allows for whatever formats to be added in the future, without any ambiguity of how to handle TimeRange.

Not sure if we support relative time at this stage, but if we do then there should be another field for what "now" was in the request. 

Not sure if the two time formats should be allowed to be different (for example, absolute from and relative to (if we support relative)). In this case, will need enum for format in both from and to. 

Thoughts @ryantxu ?

Reply:
My fear with passing strings is that each implementation would need to parse them.  I am more comfortable dropping the strings and only using ms (or nano?) -- as long as the original source for where the ms values where generated is not lost.

I vote we pull the discussion of TimeRange into its own issue since it is largely independent of the larger structural change